### DISCUSSION TITLE
Is there a way to use a different configuration for tests than for the rest?

### DISCUSSION THREAD

**Discussion by JimFawkes** [astral-sh/ruff#4119] [Q&A] [answered]
We would like to not exclude our tests directory completely however some rules make sense for "production code" but not for tests. An Example of this is S101 flagging the use of `asserts` which we obviously use in our tests but not in the "production code". We think it makes absolute sense to use this rule but just not for the tests. Which then leads us to the question:

Is there a way to use directory specific configurations of which rules to use/ignore?

**Comment by MichaReiser** [comment, marked as answer]
Ruff supports [hierarchical configurations](https://beta.ruff.rs/docs/configuration/#pyprojecttoml-discovery). This allows you to have more specific or diverging configurations per directory. Is this what you're looking for or do you have another use case?

**Comment by JimFawkes** [comment]
Thanks, yeah this works! Am I right in the assumption though that one of the two config files are used and that these would not be merged? This is not really important but just curious because with this approach I am duplicating a lot of the configuration. Although I guess this is more explicit than if we had some config here and some config there.

**Reply by charliermarsh** [reply]
That's right. Every Python file will use the configuration of the "closest" configuration file in its filesystem path.

We do support "extending" one configuration file from another, which works roughly as you've described: https://beta.ruff.rs/docs/settings/#extend. Note that the individual _fields_ aren't merged together, but it does let you inherit options from a base configuration.

**Reply by JimFawkes** [reply]
Awesome! Thank you very much for writing ruff and for taking the time to answer even though I should have found it in the docs.

**Reply by charliermarsh** [reply]
No worries! If you didn't find it in the docs, I tend to err on the side of blaming the docs :joy:
